pkg/controller/virtualmachine/virtualmachine_controller.go

```diff
@@ -81,20 +78,63 @@ type ReconcilePolicy struct {
 
 // Reconcile reads that state of the cluster for a virtual machine object and makes changes based on the state
 func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	log.V(1).Info("got a virtual machine event in the controller")
+	log.V(1).Info("got a virtual machine event in the controller", "VM", request.NamespacedName)
 	instance := &kubevirt.VirtualMachine{}
 	err := r.Get(context.TODO(), request.NamespacedName, instance)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			err := r.poolManager.ReleaseVirtualMachineMac(fmt.Sprintf("%s/%s", request.Namespace, request.Name))
-			if err != nil {
-				logging.Printf(logging.ErrorLevel, "failed to release mac for virtual machine %s: %v", request.NamespacedName, err)
-			}
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, nil
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.V(1).Info("The VM is not being deleted", "VM", request.NamespacedName)
+		// The object is not being deleted, so if it does not have a finalizer,
+		// then lets add the finalizer and update the object.
+		err = r.addFinalizerAndUpdate(instance, &request)
+		return reconcile.Result{}, err
+	}
+	// The object is being deleted
+	log.V(1).Info("The VM is being deleted", "VM", request.NamespacedName)
+	err = r.removeFinalizerAndReleaseMac(instance, &request)
+	// Our finalizer has finished, so the reconciler can do nothing.
+	return reconcile.Result{}, err
+}
+
+func (r *ReconcilePolicy) addFinalizerAndUpdate(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request) error {
+	if !helper.ContainsString(virtualMachine.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName) {
+		log.V(1).Info("The VM does not have a finalizer", "VM", request.NamespacedName)
+		virtualMachine.ObjectMeta.Finalizers = append(virtualMachine.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName)
+
+		err := r.Update(context.Background(), virtualMachine)
+		if err != nil {
+			log.Error(err, "failed to update the VM with the new finalizer", "VM", request.NamespacedName)
```

Comment:
`request.NamespacedName` is a struct please use the let's be consistent and use `VirtualMachineName` and `VirtualMachineNamespace` in all our logs